### PR TITLE
refactor(arrow2): migrate null checks to arrow-rs

### PR DESCRIPTION
**Pull request by rohitkulshreshtha**
## Summary

- Migrate `DataArray<T>::check_nullity` from arrow2 `validity()` bitmap to arrow-rs `nulls()` NullBuffer
- Migrate `PythonArray::check_nullity` to use arrow-rs `BooleanArray` directly, removing `daft_arrow::buffer::from_null_buffer` conversion
- Migrate `DataArray<T>::is_valid` from `self.data.is_valid()` (arrow2) to NullBuffer-based check
- Preserves NullArray edge case handling (null_count > 0 with no null buffer)

Note: The nested array macros (ListArray, FixedSizeListArray, StructArray) were already using arrow-rs and required no changes.

Ref: #5741

## Test plan

- [x] `cargo check -p daft-core` passes
- [x] `cargo test -p daft-core -- null` — all 51 null-related tests pass
- [x] All pre-commit hooks pass (formatting, clippy, cargo check default + all features)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/daft-core/src/array/ops/null.rs [hunk 1]
```diff
@@ -16,24 +20,32 @@ where
...
-                _ => daft_arrow::array::BooleanArray::from_slice(vec![is_null; arrow_array.len()]), // true for is_null and false for not_null
+        match self.nulls() {
+            // If there's no null buffer, check null_count for the NullArray edge case
+            None => match self.null_count() {
+                0 => Ok(BooleanArray::from((
+                    self.field.name.as_str(),
+                    repeat_n(!is_null, self.len())
+                        .collect::<Vec<_>>()
+                        .as_slice(),
+                ))),
```
Comment (universalmind303) [inline src/daft-core/src/array/ops/null.rs]: use named constructors like `from_iter` and `from_slice`. the `Array::from(tuple)` is being removed.

also never collect to a vec then create an array. Major antipattern.

File: src/daft-core/src/array/ops/null.rs [hunk 2]
```diff
@@ -16,24 +20,32 @@ where
...
+                _ => Ok(BooleanArray::from((
+                    self.field.name.as_str(),
+                    repeat_n(is_null, self.len()).collect::<Vec<_>>().as_slice(),
```
Comment (universalmind303) [inline src/daft-core/src/array/ops/null.rs]: same comment here

File: src/daft-core/src/array/ops/null.rs [hunk 3]
```diff
@@ -16,23 +20,27 @@ where
...
-            result_arrow_array,
+        let values = match self.nulls() {
+            // If there's no null buffer, check null_count for the NullArray edge case
+            None => {
+                let val = if self.null_count() == 0 {
+                    !is_null
+                } else {
+                    is_null
+                };
+                arrow::buffer::BooleanBuffer::from_iter(repeat_n(val, self.len()))
```
Comment (universalmind303) [inline src/daft-core/src/array/ops/null.rs]: perf nit: this still causes unnecessary iteration. Use `new_set`/`new_unset` instead

```rs
let all_true = BooleanBuffer::new_set(self.len());
let all_false = BooleanBuffer::new_unset(self.len());
```

**Review by universalmind303 — APPROVED** [review]
